### PR TITLE
Fix crypto subtype for trades api

### PR DESCRIPTION
**Pull request by dataCenter430**
### Problem
POST /api/v1/trades with a valid crypto account_id returns HTTP 422 with:
```
{
  "error": "validation_failed",
  "message": "Account does not support trades (investment or crypto exchange only)",
  "errors": ["Account must be an investment or crypto exchange account"]
}
```
This happens even when the account is a crypto account created in the UI and intended to be a “Crypto Exchange” (which should support trades).

### Root cause
- The Trades API only allows accounts for which account.supports_trades? is true.
- For Crypto accounts, supports_trades? is true only when the accountable’s subtype is "exchange" (see Crypto#supports_trades?).
- Crypto accounts created or edited in the UI never had subtype saved because CryptosController did not permit :subtype in accountable_attributes. The crypto form sends account[accountable_attributes][subtype], but strong parameters dropped it, so the Crypto record kept subtype NULL and supports_trades? was always false.

### Solution (plan implemented)
1. Controller
Add :subtype to the permitted accountable attributes in `app/controllers/cryptos_controller.rb` so that “Crypto Exchange” vs “Crypto Wallet” is persisted when creating or editing crypto accounts in the UI.

2. Regression tests
- In `test/controllers/cryptos_controller_test.rb`:
     - Create a crypto account via the controller with `subtype: "exchange"` and assert the record has `subtype == "exchange"` and `account.supports_trades?` is true.
   - Update an existing crypto account (with `subtype` nil) to `subtype: "exchange"` and assert the same.
- In `test/models/crypto_test.rb`: add a test that `supports_trades?` is true only when `subtype == "exchange"` and false for "wallet" and nil.

3. Verification of trade creation
Covered by the controller tests (create/update with exchange subtype and `supports_trades?`) and the new Crypto model test. No separate Trades API test file exists in the project; the fix is at the controller/permitted-params level.
4. Data backfill for existing instances
Add a migration that sets `subtype = "exchange"` for all `Crypto` records where `subtype` is NULL, so existing self-hosted (and other) instances can use the Trades API for existing crypto accounts without manual DB updates. Operators can change any account to “Crypto Wallet” in the UI if needed.

### Changes
- `app/controllers/cryptos_controller.rb`	Add :subtype to permitted_accountable_attributes (with :id, :tax_treatment).
- `test/controllers/cryptos_controller_test.rb`	Two tests: (1) create crypto with subtype "exchange" and assert persistence and supports_trades?; (2) update crypto from nil to "exchange" and assert the same.
- `test/models/crypto_test.rb`	New test: supports_trades? is true only for subtype: "exchange", false for "wallet" and nil.
- `db/migrate/20260218120000_backfill_crypto_subtype_for_trades.rb`	up: backfill Crypto.where(subtype: nil) to subtype: "exchange". down: no-op (cannot safely distinguish backfilled from user-chosen “exchange”).

### Testing
- `bin/rails test test/controllers/cryptos_controller_test.rb test/models/crypto_test.rb `, all relevant tests pass.
- Manually: create or edit a crypto account in the UI with “Crypto Exchange”, then POST `/api/v1/trades` with that account’s ID and a valid payload (e.g. positive `qty` and `price`), expect 201 Created instead of 422.

### Note for self-hosted / existing deployments
After upgrading, run `bin/rails db:migrate`. The migration will set `subtype` to `"exchange"` for all crypto accounts that currently have NULL subtype, so the Trades API works for them. If an account is actually a wallet (no trades), change its type to “Crypto Wallet” in the account settings in the UI.

### Additional
    Maybe we need to fix the UI a bit for the implementations.

## Summary by CodeRabbit

* **New Features**
  * Crypto accounts can be assigned a subtype (exchange or wallet); exchange subtypes enable trading.

* **Chores**
  * Backfilled existing crypto accounts to set subtype for correct trade behavior.
  * Controller now accepts subtype when creating/updating crypto accounts.

* **Tests**
  * Added tests ensuring subtype persistence and that only exchange subtypes enable trading.

### CODE REVIEW

File: db/migrate/20260218120000_backfill_crypto_subtype_for_trades.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+class BackfillCryptoSubtypeForTrades < ActiveRecord::Migration[7.2]
+  def up
+    # Crypto accounts created via the UI before the controller permitted :subtype
+    # had subtype NULL, so supports_trades? was false and the Trades API returned 422.
+    # Backfill to "exchange" so existing crypto accounts can use the Trades API.
+    # Operators can change subtype to "wallet" in the UI if the account is wallet-only.
+    say_with_time "Backfilling crypto subtype for existing accounts" do
+      Crypto.where(subtype: nil).update_all(subtype: "exchange")
```
[the remark this answers is not in the document]
Comment (dataCenter430) [reply db/migrate/20260218120000_backfill_crypto_subtype_for_trades.rb]: Agreed. The migration now only backfills crypto accounts that are **not linked to any provider** (no row in `account_providers`). So:

- **Manual** crypto accounts (created in the UI, no CoinStats/other link) with `subtype` NULL get `subtype = 'exchange'` and can use the Trades API.
- **Wallet-linked** accounts (e.g. CoinStats wallet via `CoinstatsItem::WalletLinker#create_account_from_token`, which creates `Crypto.new` without a subtype) stay with `subtype` NULL and remain wallet/sync-only; `supports_trades?` stays false.

The backfill uses a raw SQL `UPDATE` that restricts to cryptos whose account has no `account_providers` row, so we never relabel provider-linked wallets as exchanges.

**Comment by dataCenter430** [comment]
> 🧹 Nitpick comments (2)
> > db/migrate/20260218120000_backfill_crypto_subtype_for_trades.rb (1)> `10-10`: **Avoid referencing the `Crypto` AR model class directly in a migration.**
> > > Using `Crypto.where(...).update_all(...)` couples this migration to the current state of the `Crypto` model. If the model is renamed, deleted, or later gains incompatible hooks, running `db:migrate` from scratch will fail at this point. The safe idiomatic approach is raw SQL via `execute`:
> > > 
> > > ♻️ Proposed refactor using raw SQL
> > > ```diff
> > > -      Crypto.where(subtype: nil).update_all(subtype: "exchange")
> > > +      execute "UPDATE cryptos SET subtype = 'exchange' WHERE subtype IS NULL"
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 
> > > 
> > > 🤖 Prompt for AI Agents
> > > ```
> > > Verify each finding against the current code and only fix it if needed.
> > > 
> > > In `@db/migrate/20260218120000_backfill_crypto_subtype_for_trades.rb` at line 10,
> > > The migration currently calls the ActiveRecord model call Crypto.where(subtype:
> > > nil).update_all(subtype: "exchange"), which ties the migration to the runtime
> > > model; replace this with a raw SQL update using execute to avoid coupling to the
> > > Crypto model (e.g. an UPDATE cryptos SET subtype = 'exchange' WHERE subtype IS
> > > NULL executed via execute). Ensure you reference the actual table name (cryptos)
> > > and use SQL NULL checks so the migration remains robust if the Crypto model
> > > changes or is removed.
> > > ```
> > 
> > 
> > 
> > test/controllers/cryptos_controller_test.rb (1)> `12-31`: **LGTM — minor: consider a more deterministic way to find the created account.**
> > > The `order(:created_at).last` approach is fine in practice (PostgreSQL timestamps have microsecond precision), but a more robust pattern avoids timestamp ordering entirely by parsing the redirect location:
> > > 
> > > ♻️ More deterministic record lookup
> > > ```diff
> > > -    created = `@family.accounts.where`(accountable_type: "Crypto").order(:created_at).last
> > > -    assert_redirected_to created
> > > +    assert_response :redirect
> > > +    created = Account.find(response.location[/\d+$/])
> > >      assert_equal "exchange", created.accountable.subtype, "subtype must be persisted for trades API"
> > >      assert created.supports_trades?, "exchange crypto account must support trades"
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Or equivalently, keep `assert_difference` as-is and replace the lookup with:
> > > ```diff
> > > -    created = `@family.accounts.where`(accountable_type: "Crypto").order(:created_at).last
> > > +    created = `@family.accounts.order`(:created_at).last
> > >      assert_redirected_to created
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 
> > > 
> > > 🤖 Prompt for AI Agents
> > > ```
> > > Verify each finding against the current code and only fix it if needed.
> > > 
> > > In `@test/controllers/cryptos_controller_test.rb` around lines 12 - 31, Replace
> > > the timestamp-based lookup with a deterministic parse of the redirect location:
> > > after the post to cryptos_path, grab the redirect URL from
> > > response.headers["Location"] (or response.location), extract the created account
> > > id from the path, then load the record via `@family.accounts.find`(id) (or
> > > Account.find(id)) and assign it to created; keep the subsequent assertions
> > > (assert_redirected_to created, assert_equal created.accountable.subtype,
> > > created.supports_trades?) unchanged so the test no longer relies on created_at
> > > ordering.
> > > ```
> 
> 🤖 Prompt for all review comments with AI agents
> ```
> Verify each finding against the current code and only fix it if needed.
> 
> Nitpick comments:
> In `@db/migrate/20260218120000_backfill_crypto_subtype_for_trades.rb`:
> - Line 10: The migration currently calls the ActiveRecord model call
> Crypto.where(subtype: nil).update_all(subtype: "exchange"), which ties the
> migration to the runtime model; replace this with a raw SQL update using execute
> to avoid coupling to the Crypto model (e.g. an UPDATE cryptos SET subtype =
> 'exchange' WHERE subtype IS NULL executed via execute). Ensure you reference the
> actual table name (cryptos) and use SQL NULL checks so the migration remains
> robust if the Crypto model changes or is removed.
> 
> In `@test/controllers/cryptos_controller_test.rb`:
> - Around line 12-31: Replace the timestamp-based lookup with a deterministic
> parse of the redirect location: after the post to cryptos_path, grab the
> redirect URL from response.headers["Location"] (or response.location), extract
> the created account id from the path, then load the record via
> `@family.accounts.find`(id) (or Account.find(id)) and assign it to created; keep
> the subsequent assertions (assert_redirected_to created, assert_equal
> created.accountable.subtype, created.supports_trades?) unchanged so the test no
> longer relies on created_at ordering.
> ```

The migration no longer uses `Crypto.where(...).update_all(...)`. It now uses a single `execute` with raw SQL:
```sql
UPDATE cryptos
SET subtype = 'exchange'
WHERE subtype IS NULL
  AND id IN (
    SELECT a.accountable_id
    FROM accounts a
    WHERE a.accountable_type = 'Crypto'
      AND NOT EXISTS (SELECT 1 FROM account_providers ap WHERE ap.account_id = a.id)
  )
```
So the migration no longer depends on the `Crypto` model and stays valid if the model is renamed or changed.

The test no longer uses `order(:created_at).last`. It now:
1. Asserts `assert_response :redirect`.
2. Parses the redirect URL with `URI(response.location).path.split("/").last` to get the created account ID (UUID).
3. Loads the record with `Account.find(...)` and keeps the existing assertions.
So the created account is identified from the redirect location instead of timestamp ordering.

**Comment by dataCenter430** [comment]
Change: The controller test now:
- Uses `assert_response :redirect`
- Gets the created account with `Account.find(URI(response.location).path.split("/").last) `(works with UUIDs)
- Keeps the same assertions
`bin/rails test test/controllers/cryptos_controller_test.rb` passes

**Comment by dataCenter430** [comment]
Hi, @jjmata  Could you please review the PR? I hope your suggestion with this.
Thank you.

**Comment by jjmata** [comment]
Can you make sure tests pass first @dataCenter430?  Will review as soon as that is the case, otherwise I can't merge if all else is good.

**Comment by dataCenter430** [comment]
Hi, @jjmata  I'm sure I passed tests fully.
so expected response:

`http://localhost:3000/api/v1/trades POST`

Headers:
`Content-Type: application/json`,
`X-Api-Key: API_KEY`

Raw: 
```
{
       "trade": {
         "type": "sell",
         "qty": 0.0135,
         "price": 150,
         "date": "2026-02-18",
         "account_id": "d2f27731-5ede-456c-a4f7-ae816c99cf43",
         "ticker": "SOL-USD"
       }
     }
```
Response(201):

```
{
    "id": "ca6d0b84-511b-4b2d-ad27-976f00e61afa",
    "date": "2026-02-18",
    "amount": "-$2.03",
    "currency": "USD",
    "name": "Sell 0.0135 shares of SOL-USD",
    "notes": null,
    "qty": "-0.0135",
    "price": "$150.00",
    "investment_activity_label": "Sell",
    "account": {
        "id": "d2f27731-5ede-456c-a4f7-ae816c99cf43",
        "name": "Maxim",
        "account_type": "crypto"
    },
    "security": {
        "id": "20af9a73-a127-4b38-8cc4-514351e2d8a3",
        "ticker": "SOL-USD",
        "name": null
    },
    "category": null,
    "created_at": "2026-02-19T11:21:58Z",
    "updated_at": "2026-02-19T11:21:58Z"
}
```

**Comment by jjmata** [comment]
> Hi, @jjmata I'm sure I passed tests fully.

The CI is not even getting to the tests because you have a DB migration problem.  Check GitHub: [ci / test (pull_request)](https://github.com/we-promise/sure/actions/runs/22168593335/job/64101427853?pr=1022)Failing after 57s

**Comment by dataCenter430** [comment]
Hi, @jjmata  thank you catching this out,
I wasn't pushed the schema updated at previous time, sorry for the mistake.
could you review it again?
Best regards
